diff --git a/hints/linux.sh b/hints/linux.sh
index 83ba0c5..d5a7f25 100644
--- a/hints/linux.sh
+++ b/hints/linux.sh
@@ -486,3 +486,10 @@ case "$libdb_needs_pthread" in
     libswanted="$libswanted pthread"
     ;;
 esac
+
+# Detect case-insensitive file systems
+echo X >UU/casesense.tmp
+if test -f UU/CASESENSE.TMP && test -f UU/CaSeSeNsE.tMp; then
+    firstmakefile='GNUmakefile'
+fi
+rm -f UU/casesense.tmp
