From ccf3691580f13054799caa66d9165b0c98178918 Mon Sep 17 00:00:00 2001
From: Peter Martini <PeterCMartini@GMail.com>
Date: Wed, 15 Apr 2015 22:08:17 -0400
Subject: [PATCH] tmp

---
 hints/linux.sh | 19 ++++++++++++++++++-
 patchlevel.h   |  1 +
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/hints/linux.sh b/hints/linux.sh
index 956adfc..fb5a46e 100644
--- a/hints/linux.sh
+++ b/hints/linux.sh
@@ -178,6 +178,23 @@ case "$plibpth" in
     ;;
 esac
 
+# libquadmath is sometimes installed as gcc internal library,
+# so contrary to our usual policy of *not* looking at gcc internal
+# directories we now *do* look at them, in case they contain
+# the quadmath library.
+# XXX This may apply to other gcc internal libraries, if such exist.
+# XXX This could be at Configure level, but then the $gcc is messy.
+case "$usequadmath" in
+"$define")
+  for d in `LANG=C LC_ALL=C $gcc $ccflags $ldflags -print-search-dirs | grep libraries | cut -f2- -d= | tr ':' $trnl | grep 'gcc' | sed -e 's:/$::'`
+  do
+    case `ls $d/*libquadmath*$so* 2>/dev/null` in
+    $d/*libquadmath*$so*) xlibpth="$xlibpth $d" ;;
+    esac
+  done
+  ;;
+esac
+
 case "$libc" in
 '')
 # If you have glibc, then report the version for ./myconfig bug reporting.
@@ -497,7 +514,7 @@ then
        DBLIB="$DBDIR/libdb.so"
        if [ -f $DBLIB ]
        then
-         if ${nm:-nm} -u $DBLIB | grep pthread >/dev/null
+         if ${nm:-nm} -u $DBLIB 2>/dev/null | grep pthread >/dev/null
          then
            if ldd $DBLIB | grep pthread >/dev/null
            then
diff --git a/patchlevel.h b/patchlevel.h
index 07e1217..8016ec8 100644
--- a/patchlevel.h
+++ b/patchlevel.h
@@ -137,6 +137,7 @@ static const char * const local_patches[] = {
 	,"uncommitted-changes"
 #endif
 	PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
+	,"Devel::PatchPerl 1.32"
 	,NULL
 };
 
-- 
1.9.1

